### ISSUE TITLE
Sometimes stopping an app stops two or more apps

### ISSUE THREAD

**Issue by miyconst** [Polyjuice/AppStoreClient#7]
From time to time, when I stop an app some other apps got stopped too.

I tried to reproduce the issue to find what causes this, but was not able to find a 100% sure configuration when stopping an app would also stop another one.

Most of the time this happens when those apps were started almost at the same time, but not always like this.

**Comment by miyconst** [comment]
Steps to reproduce:

1. Open a Visual Studio instance.
2. Open a project (for example Products) in the Visual Studio instance.
3. Run the project with `Ctrl + F5`.
4. Close the project.
5. Open another project (for example People) in the Visual Studio instance.
6. Run the project with `Ctrl + F5`.
7. Go to administration panel at `http://localhost:8181/`.
8. Stop the first project (Product).
9. The People application will also be stopped.

This also does not let me to run multiply projects from one instance of Visual Studio.

cc @chrhol, @alemoi

**Comment by alemoi** [comment]
Hmm I'm curious how it worked before. I would say we should consult with Per who is not in the list of users for Polyjuice here though.

**Comment by miyconst** [comment]
It is like this from the very beginning, and I thought it is a bug of AppStoreClient, but looks like it is not.

**Comment by miyconst** [comment]
@per-samuelsson, could you provide any input on this?

**Comment by per-samuelsson** [comment]
> @per-samuelsson, could you provide any input on this?

Yes, its kind of by design, but a design that is not optimal. Its a bit tricky to get this whole debugging experience 100% crisp, given the model we have to run apps in a shared process.

I can test it at some point and see if I can refine it more, but right now I'm pretty choked with other stuff I have to do. Can we make it like "semi-prio"? :smile:

**Comment by miyconst** [comment]
Yes, sure, I was about to close the issue, as it is not that important.

**Comment by warpech** [comment]
I just want to confirm that I find it annoying as well :) 

But I understand that it is not a priority and may be gone when we switch to run apps in separate processes?

**Comment by dan31** [comment]
I think, first of all, for current 2.x generation, it is good if it is explained, in which circumstance all apps are to be stopped by Starcounter and why. From @miyconst post in the header it appears that it doesn't always stop all the apps when one stopped, but only time to time. Hence, good to know would be, when this happens and when is it not? @per-samuelsson, is there some simple answer?

@warpech, for sure, in 3.x with multi-process architecture stopping/starting apps will be in a blink of an eye. One of many benefits of this new architecture. Currently, the main reason for this behaviors is because SC hosts its apps in the runtime which it doesn't own (in contrast to SAP HANA, where SAP is also a language runtime).

**Comment by per-samuelsson** [comment]
> I think, first of all, for current 2.x generation, it is good if it is explained, in which circumstance all apps are to be stopped by Starcounter and why. From @miyconst post in the header it appears that it doesn't always stop all the apps when one stopped, but only time to time. Hence, good to know would be, when this happens and when is it not? @per-samuelsson, is there some simple answer?

I totally agree. But the problem here is that the design itself is bad IMO, I've thought so from the first day I actually did it (it's my own), but I just can't find any other way to do it, considering how VS work, attaching debugger work and how that relate to our single-host, multi-app design.

I hinted about already in my earlier comments.

I really don't have any big hopes (because I've been thinking on alternatives on and off for quite some time), but my aim this week is to give this thing a new round of thinking and see if I can come up with something. I really don't want to document it before that. And users will likely find it quite awkward too, because it is. :smile:

**Comment by per-samuelsson** [comment]
> @warpech, for sure, in 3.x with multi-process architecture stopping/starting apps will be in a blink of an eye. One of many benefits of this new architecture. 

Not the entire truth I'm afraid, since we've said that we still need to support multi-app hosting processes, and then our options are basically the same as today.

However, maybe we can design it a little bit differently, and kind of say that on development environments, processes are always 1-host-per-app, and that multi-app hosts are only for production (to save resources). Then we could improve the cycle (start-stop-restart) a **lot** IMO.

**Comment by miyconst** [comment]
Just a short note to remove the confusion.

The behavior is not random, it is static: Stopping an app stops all of the other apps started from same Visual Studio instance.

If two apps started from two different Visual Studio instances, then stopping one does not affect another once.

I have changed the title.

**Comment by dan31** [comment]
@miyconst that is good explanation! Thanks for clarification, I think that makes the behavior explained as predictable.
@per-samuelsson To be honest, you have a complex design task solved, thus I won't say it is a bad design, since it is constrained by many factors we don't influence. It is perfectly fine to change architectures in future generations. And I see your current hopes, maybe you'd really manage to make it better through your current redesign, which was made necessary because of the users' feedback.

**Comment by per-samuelsson** [comment]
> The behavior is not random, it is static: Stopping an app stops all of the other apps started from same Visual Studio instance.

Yes, it's static because it's by design.

> @per-samuelsson To be honest, you have a complex design task solved, thus I won't say it is a bad design, since it is constrained by many factors we don't influence. It is perfectly fine to change architectures in future generations. And I see your current hopes, maybe you'd really manage to make it better through your current redesign, which was made necessary because of the users' feedback.

That's nice of you, @dan31. :beers: I'd still like us to review this thingy, and my aim is to do that during this week. Let's see if I can come up with something. And/or then document it somehow.

**Comment by per-samuelsson** [comment]
> The behavior is not random, it is static: Stopping an app stops all of the other apps started from same Visual Studio instance.

While I'm at it, let me try to give a short insight in why the underlying design is like it is.

Picture a solution, "MyApps.sln", with 3 Starcounter application projects. You configure that when F5 is hit, VS should "Start project" for all 3 apps.

On the other end, we support **stopping** of applications by the only possible way:

1. We stop the host (i.e. unload all applications that run in it)
2. We restart the host (empty)
3. We restart every app (i.e. reload the exe) that was running when **Stop application** fired, all applications **but** the one that is now stopped.

Back to "MyApps.sln". Every time the developer hit F5, what should happen?

1. All projects should build.
2. Every project that is configured to start should start.

Lets do **staradmin kill all** to start fresh. After doing that, we do F5 on "MyApps.sln". Now, in Visual Studio, projects are *isolated*, i.e. they don't know about each other. Hence, every project will be asked to "start itself", and that's when our extension kick in - we take over what it means to start it.

The first project will find that the server does not run (because of previous **staradmin kill all**), and neither the host. It will start those, and then load itself into the empty host. The next project will find that the host is running, so it will check "Am I already running inside that host?". It's not. Hence it can just load itself into that host. And the same steps will happen for the third project.

Now, the state we have is that the code host is running with 3 applications in it - those under the control of "MyApps.sln".

Hit stop. Then hit F5 again. What should happen? We should reload all the 3 applications, because they might have changed.

The first application will get the call. It will see that a) the code host is running, and b) that **an old version of itself is running inside that host**. What do do? Stop the code host (and hence all applications in it, by necessity), and reload itself into the code host after it has been restarted.

Now, when that happens: should it also **restart all other applications**? Remember: the two other applications (that will soon recieve the F5 command) are loaded in the code host that now has stopped. Let's say it does: it restart those two, and then load itself (the new version) into the host. The result: the host with 3 applications.

The comes F5 to the second project: this project will find the state above; a running host, and an old version of itself running in it. What do do? Stop the host, restart the host. Reload project 1, and then insert itself into the host.

And then the same for third project (or forth, or fifth, or 12th like in Tunity case). For every project, the code host must restart. 3 times for 3 projects, or 12 for 12. And for every time, we have to again **attach the debugger** (since we constantly have to recycle the host process).

Another parameter is that if we do like that, what would happen if you first start "MyApps.sln" with 3 projects configured to be started. Run it. And then stop and configure just 2 to run. When you hit F5, all three applications would still run, because of that first project that would reload any application already running inside the host when it's asked to start.

Hence, the design when restarting must be somewhat differently from VS - that's why we kind of have to double the logic that make up `star.exe`. If not for this, VS could simply invoke `star.exe` entrypoint. Now it can't. Instead, it has to keep track of restarts so that every time you restart a **configured set of projects** (which is actually what F5 does), every project in that set has to be responsible to restart itself. No one else will. And hence it can not work exactly as we would like it too, and the way it works outside the context of VS.
